cub/examples/device/example_device_topk_keys.cu

```diff
@@ -0,0 +1,191 @@
+ * Simple example of cub::DeviceTopK::MinKeys().
+ *
+ * Find the top-k smallest float keys paired with a corresponding array of int values.
+ *
+ * To compile using the command line:
+ *   nvcc -arch=sm_XX example_device_topk_keys.cu -I../.. -lcudart -O3
+ *
+ ******************************************************************************/
+
+// Ensure printing of CUDA runtime errors to console
+#define CUB_STDERR
+
+#include <cub/device/device_topk.cuh>
+#include <cub/util_allocator.cuh>
+
+#include <algorithm>
+
+#include <stdio.h>
+
+#include "../../test/test_util.h"
+
+using namespace cub;
+
+//---------------------------------------------------------------------
+// Globals, constants and aliases
+//---------------------------------------------------------------------
+
+bool g_verbose = false; // Whether to display input/output to console
+CachingDeviceAllocator g_allocator(true); // Caching allocator for device memory
+
+//---------------------------------------------------------------------
+// Test generation
+//---------------------------------------------------------------------
+/**
+ * Initialize key-value sorting problem.
+ */
+
+void Initialize(float* h_keys, float* h_reference_keys, int num_items, int k)
+{
+  for (int i = 0; i < num_items; ++i)
+  {
+    RandomBits(h_keys[i]);
+  }
+
+  if (g_verbose)
+  {
+    printf("Input keys:\n");
+    DisplayResults(h_keys, num_items);
+    printf("\n\n");
+  }
+
+  std::partial_sort_copy(h_keys, h_keys + num_items, h_reference_keys, h_reference_keys + k);
+}
+/**
+ *  In some case the results of topK is unordered. Sort the results to compare with groundtruth.
+ */
+void SortUnorderedRes(float* h_res_keys, float* d_keys_out, int k)
+{
+  CubDebugExit(cudaMemcpy(h_res_keys, d_keys_out, sizeof(float) * k, cudaMemcpyDeviceToHost));
+  std::stable_sort(h_res_keys, h_res_keys + k);
+}
+
+//---------------------------------------------------------------------
+// Main
+//---------------------------------------------------------------------
+
+/**
+ * Main
+ */
+int main(int argc, char** argv)
+{
+  int num_items = 10240;
+  int k         = 10;
+  // Initialize command line
+  CommandLineArgs args(argc, argv);
+  g_verbose = args.CheckCmdLineFlag("v");
+  args.GetCmdLineArgument("n", num_items);
+  args.GetCmdLineArgument("k", k);
+  // Print usage
+  if (args.CheckCmdLineFlag("help"))
+  {
+    printf("%s "
+           "[--n=<input items> "
+           "[--k=<output items> "
+           "[--device=<device-id>] "
+           "[--v] "
+           "\n",
+           argv[0]);
+    exit(0);
+  }
+
+  // Initialize device
+  CubDebugExit(args.DeviceInit());
+
+  printf("cub::DeviceTopK::MinKeys() find %d smallest items from %d items (%d-byte keys)\n",
+         k,
+         num_items,
+         int(sizeof(float)));
+  fflush(stdout);
+
+  // Allocate host arrays
+  float* h_keys           = new float[num_items];
+  float* h_reference_keys = new float[k];
+  float* h_res_keys       = new float[k];
```

Comment:
Important: Our examples should use best practices for C++ and thus avoid naked allocations. Please use a `std::vector` or a `std::unique_ptr`. Or a `thrust::host_vector`.